src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -428,24 +433,47 @@ public void saveProgram(File f) throws IOException {
             return;
         }
 
-        try {
-            RiverWriter w = new RiverWriter(tmpFile, execution.getOwner());
+        for (int i = 0; i < MAX_RETRIES; ++i) {
             try {
-                w.writeObject(this);
+                RiverWriter w = new RiverWriter(tmpFile, execution.getOwner());
+                try {
+                    w.writeObject(this);
+                } finally {
+                    w.close();
+                }
+                Files.move(tmpFile.toPath(), f.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
+                LOGGER.log(FINE, "program state saved");
+                break;
+            } catch (RuntimeException e) {
+                if (i < MAX_RETRIES - 1) {
+                    LOGGER.log(FINE, "Retry [" + (i + 1) + "/" + MAX_RETRIES + "]", e);
+                    long timeout = Math.min(1000, (long) (10 * Math.pow(2, i)));
+                    try {
+                        TimeUnit.MILLISECONDS.sleep(timeout);
+                        continue;
+                    } catch (InterruptedException ignore) {
+                        // give up
+                    }
+                }
+                propagateErrorToWorkflow(e);
+                throw new IOException("Failed to persist "+f,e);
+            } catch (IOException e) {
+                if (i < MAX_RETRIES - 1) {
+                    LOGGER.log(FINE, "Retry [" + (i + 1) + "/" + MAX_RETRIES + "]", e);
+                    long timeout = Math.min(1000, (long) (10 * Math.pow(2, i)));
+                    try {
+                        TimeUnit.MILLISECONDS.sleep(timeout);
```

Comment:
@jglick I understand the first request (catch `move` but not `writeObject`) but I don't fully understand the second request.

I see how `CpsStepContext.saveState` invokes `saveProgram()` from inside `runInCpsVmThread()`, but what are you asking me to do about this in the context of `CpsThreadGroup.saveProgram()`?